File: web/scada-web/src/visual-runtime/builtinVisualObjectSchemas.ts
import {
  VISUAL_PROPERTY_KEYS,
  VisualObjectPropertySchema,
  type CommonVisualPropertyKey
} from './visualPropertyRegistry';

export const BUILTIN_VISUAL_OBJECT_TYPES = {
  group: 'core.group',
  rectangle: 'core.rectangle',
  ellipse: 'core.ellipse',
  line: 'core.line',
  polygon: 'core.polygon',
  text: 'core.text',
  image: 'core.image',
  valueDisplay: 'core.valueDisplay',
  trend: 'core.trend',
  alarmBrowser: 'core.alarmBrowser',
  eventBrowser: 'core.eventBrowser',
  button: 'core.button',
  slider: 'core.slider'
} as const;

export type BuiltinVisualObjectType = typeof BUILTIN_VISUAL_OBJECT_TYPES[keyof typeof BUILTIN_VISUAL_OBJECT_TYPES];

const ANALOG_FILL_CAPABLE_TYPES = new Set<string>([
  BUILTIN_VISUAL_OBJECT_TYPES.rectangle,
  BUILTIN_VISUAL_OBJECT_TYPES.ellipse
]);

const GEOMETRY: readonly CommonVisualPropertyKey[] = [
  VISUAL_PROPERTY_KEYS.x,
  VISUAL_PROPERTY_KEYS.y,
  VISUAL_PROPERTY_KEYS.width,
  VISUAL_PROPERTY_KEYS.height,
  VISUAL_PROPERTY_KEYS.zIndex
];

const TRANSFORM: readonly CommonVisualPropertyKey[] = [
  VISUAL_PROPERTY_KEYS.rotation,
  VISUAL_PROPERTY_KEYS.scaleX,
  VISUAL_PROPERTY_KEYS.scaleY,
  VISUAL_PROPERTY_KEYS.horizontalFlip,
  VISUAL_PROPERTY_KEYS.verticalFlip
];

const VISIBILITY: readonly CommonVisualPropertyKey[] = [
  VISUAL_PROPERTY_KEYS.visible,
  VISUAL_PROPERTY_KEYS.opacity,
  VISUAL_PROPERTY_KEYS.tooltip,
  VISUAL_PROPERTY_KEYS.enabled
];

const EFFECTS: readonly CommonVisualPropertyKey[] = [
  VISUAL_PROPERTY_KEYS.shadowEnabled,
  VISUAL_PROPERTY_KEYS.shadowColor,
  VISUAL_PROPERTY_KEYS.shadowOffsetX,
  VISUAL_PROPERTY_KEYS.shadowOffsetY,
  VISUAL_PROPERTY_KEYS.shadowBlur
];

const FILL: readonly CommonVisualPropertyKey[] = [
  VISUAL_PROPERTY_KEYS.fillStyle,
  VISUAL_PROPERTY_KEYS.fillColor,
  VISUAL_PROPERTY_KEYS.fillSecondaryColor,
  VISUAL_PROPERTY_KEYS.gradientDirection
];

const STROKE: readonly CommonVisualPropertyKey[] = [
  VISUAL_PROPERTY_KEYS.strokeColor,
  VISUAL_PROPERTY_KEYS.strokeWidth,
  VISUAL_PROPERTY_KEYS.strokeStyle
];

const TEXT: readonly CommonVisualPropertyKey[] = [
  VISUAL_PROPERTY_KEYS.text,
  VISUAL_PROPERTY_KEYS.textColor,
  VISUAL_PROPERTY_KEYS.fontFamily,
  VISUAL_PROPERTY_KEYS.fontSize,
  VISUAL_PROPERTY_KEYS.fontWeight,
  VISUAL_PROPERTY_KEYS.fontStyle,
  VISUAL_PROPERTY_KEYS.underline,
  VISUAL_PROPERTY_KEYS.textWrap,
  VISUAL_PROPERTY_KEYS.lineHeight,
  VISUAL_PROPERTY_KEYS.textOverflow,
  VISUAL_PROPERTY_KEYS.horizontalAlignment,
  VISUAL_PROPERTY_KEYS.verticalAlignment
];

const TREND: readonly CommonVisualPropertyKey[] = [
  VISUAL_PROPERTY_KEYS.backgroundColor,
  VISUAL_PROPERTY_KEYS.strokeColor,
  VISUAL_PROPERTY_KEYS.strokeWidth,
  VISUAL_PROPERTY_KEYS.cornerRadius,
  VISUAL_PROPERTY_KEYS.trendMode,
  VISUAL_PROPERTY_KEYS.trendWindowSeconds,
  VISUAL_PROPERTY_KEYS.trendRefreshSeconds,
  VISUAL_PROPERTY_KEYS.trendLegendVisible,
  VISUAL_PROPERTY_KEYS.trendGridVisible,
  VISUAL_PROPERTY_KEYS.trendAxesVisible,
  VISUAL_PROPERTY_KEYS.trendQualityVisible
];

const BROWSER: readonly CommonVisualPropertyKey[] = [
  VISUAL_PROPERTY_KEYS.backgroundColor,
  VISUAL_PROPERTY_KEYS.strokeColor,
  VISUAL_PROPERTY_KEYS.strokeWidth,
  VISUAL_PROPERTY_KEYS.cornerRadius
];

const BASE = [...GEOMETRY, ...TRANSFORM, ...VISIBILITY, ...EFFECTS] as const;

const schemas = new Map<BuiltinVisualObjectType, VisualObjectPropertySchema>([
  [BUILTIN_VISUAL_OBJECT_TYPES.group, schema(BUILTIN_VISUAL_OBJECT_TYPES.group, BASE)],
  [BUILTIN_VISUAL_OBJECT_TYPES.rectangle, schema(BUILTIN_VISUAL_OBJECT_TYPES.rectangle, [
    ...BASE,
    ...FILL,
    ...STROKE,
    VISUAL_PROPERTY_KEYS.cornerRadius
  ])],
  [BUILTIN_VISUAL_OBJECT_TYPES.ellipse, schema(BUILTIN_VISUAL_OBJECT_TYPES.ellipse, [
    ...BASE,
    ...FILL,
    ...STROKE
  ])],
  [BUILTIN_VISUAL_OBJECT_TYPES.line, schema(BUILTIN_VISUAL_OBJECT_TYPES.line, [
    ...BASE,
    ...STROKE
  ])],
  [BUILTIN_VISUAL_OBJECT_TYPES.polygon, schema(BUILTIN_VISUAL_OBJECT_TYPES.polygon, [
    ...BASE,
    ...FILL,
    ...STROKE
  ])],
  [BUILTIN_VISUAL_OBJECT_TYPES.text, schema(BUILTIN_VISUAL_OBJECT_TYPES.text, [
    ...BASE,
    ...TEXT
  ])],
  [BUILTIN_VISUAL_OBJECT_TYPES.image, schema(BUILTIN_VISUAL_OBJECT_TYPES.image, [
    ...BASE,
    VISUAL_PROPERTY_KEYS.assetRef,
    VISUAL_PROPERTY_KEYS.imageFit,
    VISUAL_PROPERTY_KEYS.imagePositionX,
    VISUAL_PROPERTY_KEYS.imagePositionY
  ])],
  [BUILTIN_VISUAL_OBJECT_TYPES.valueDisplay, schema(BUILTIN_VISUAL_OBJECT_TYPES.valueDisplay, [
    ...BASE,
    VISUAL_PROPERTY_KEYS.backgroundColor,
    ...STROKE,
    VISUAL_PROPERTY_KEYS.cornerRadius,
    ...TEXT
  ])],
  [BUILTIN_VISUAL_OBJECT_TYPES.trend, schema(BUILTIN_VISUAL_OBJECT_TYPES.trend, [
    ...BASE,
    ...TREND
  ])],
  [BUILTIN_VISUAL_OBJECT_TYPES.alarmBrowser, schema(BUILTIN_VISUAL_OBJECT_TYPES.alarmBrowser, [
    ...BASE,
    ...BROWSER
  ])],
  [BUILTIN_VISUAL_OBJECT_TYPES.eventBrowser, schema(BUILTIN_VISUAL_OBJECT_TYPES.eventBrowser, [
    ...BASE,
    ...BROWSER
  ])],
  [BUILTIN_VISUAL_OBJECT_TYPES.button, schema(BUILTIN_VISUAL_OBJECT_TYPES.button, [
    ...BASE,
    VISUAL_PROPERTY_KEYS.backgroundColor,
    ...STROKE,
    VISUAL_PROPERTY_KEYS.cornerRadius,
    ...TEXT
  ])],
  [BUILTIN_VISUAL_OBJECT_TYPES.slider, schema(BUILTIN_VISUAL_OBJECT_TYPES.slider, [
    ...BASE,
    VISUAL_PROPERTY_KEYS.value,
    VISUAL_PROPERTY_KEYS.minimum,
    VISUAL_PROPERTY_KEYS.maximum,
    VISUAL_PROPERTY_KEYS.step,
    VISUAL_PROPERTY_KEYS.orientation,
    VISUAL_PROPERTY_KEYS.interactionEnabled,
    VISUAL_PROPERTY_KEYS.reverseDirection,
    VISUAL_PROPERTY_KEYS.trackColor,
    VISUAL_PROPERTY_KEYS.thumbColor,
    VISUAL_PROPERTY_KEYS.strokeColor,
    VISUAL_PROPERTY_KEYS.strokeWidth,
    VISUAL_PROPERTY_KEYS.cornerRadius
  ])]
]);

export function supportsAnalogFill(objectType: string): boolean {
  return ANALOG_FILL_CAPABLE_TYPES.has(objectType);
}

export function getBuiltinVisualObjectSchema(objectType: string): VisualObjectPropertySchema {
  const result = schemas.get(objectType as BuiltinVisualObjectType);
  if (!result) {
    throw new Error(`Unknown built-in visual object type '${objectType}'.`);
  }
  return result;
}

export function listBuiltinVisualObjectSchemas(): readonly VisualObjectPropertySchema[] {
  return Object.freeze([...schemas.values()]);
}

function schema(
  objectType: BuiltinVisualObjectType,
  propertyKeys: readonly CommonVisualPropertyKey[]
): VisualObjectPropertySchema {
  return new VisualObjectPropertySchema(objectType, propertyKeys);
}

File: src/Scada.Engineering/VisualScripting/BuiltinVisualObjectSchemas.cs
namespace Scada.Engineering.VisualScripting;

/// <summary>
/// Renderer-independent built-in object schemas shared by Engineering, Runtime
/// and the future graphical editor. This is a type/property contract only; it
/// does not implement a canvas, palette UI or renderer.
/// </summary>
public static class BuiltinVisualObjectSchemas
{
    public const string GroupType = "core.group";
    public const string RectangleType = "core.rectangle";
    public const string EllipseType = "core.ellipse";
    public const string LineType = "core.line";
    public const string PolygonType = "core.polygon";
    public const string TextType = "core.text";
    public const string ImageType = "core.image";
    public const string ValueDisplayType = "core.valueDisplay";
    public const string TrendType = "core.trend";
    public const string ButtonType = "core.button";
    public const string SliderType = "core.slider";
    public const string TrendPensProperty = "pens";

    private const string TrendModeProperty = "trendMode";
    private const string TrendWindowSecondsProperty = "trendWindowSeconds";
    private const string TrendRefreshSecondsProperty = "trendRefreshSeconds";
    private const string TrendLegendVisibleProperty = "trendLegendVisible";
    private const string TrendGridVisibleProperty = "trendGridVisible";
    private const string TrendAxesVisibleProperty = "trendAxesVisible";
    private const string TrendQualityVisibleProperty = "trendQualityVisible";

    private static readonly IReadOnlyList<VisualPropertyDefinition> TrendDefinitions =
    [
        new(
            TrendModeProperty,
            new VisualStringValue("history"),
            constraints: new VisualPropertyConstraints
            {
                AllowedValues = ["history", "live"],
                AllowEmptyString = false
            }),
        new(
            TrendWindowSecondsProperty,
            new VisualIntegerValue(3600),
            constraints: new VisualPropertyConstraints { Minimum = 60, Maximum = 604800 },
            unit: "s"),
        new(
            TrendRefreshSecondsProperty,
            new VisualIntegerValue(5),
            constraints: new VisualPropertyConstraints { Minimum = 1, Maximum = 3600 },
            unit: "s"),
        new(TrendLegendVisibleProperty, new VisualBooleanValue(true)),
        new(TrendGridVisibleProperty, new VisualBooleanValue(true)),
        new(TrendAxesVisibleProperty, new VisualBooleanValue(true)),
        new(TrendQualityVisibleProperty, new VisualBooleanValue(true))
    ];

    private static readonly IReadOnlyDictionary<string, VisualPropertyDefinition> CommonByKey =
        CommonVisualPropertyDefinitions.Geometry
            .Concat(CommonVisualPropertyDefinitions.Transform)
            .Concat(CommonVisualPropertyDefinitions.Visibility)
            .Concat(CommonVisualPropertyDefinitions.Fill)
            .Concat(CommonVisualPropertyDefinitions.Stroke)
            .Concat(CommonVisualPropertyDefinitions.Effects)
            .Concat(CommonVisualPropertyDefinitions.Text)
            .Concat(CommonVisualPropertyDefinitions.Image)
            .Concat(CommonVisualPropertyDefinitions.Slider)
            .Concat(TrendDefinitions)
            .ToDictionary(property => property.Key, StringComparer.Ordinal);

    private static readonly HashSet<string> AnalogFillCapableTypes =
        new(StringComparer.Ordinal)
        {
            RectangleType,
            EllipseType
        };

    private static readonly string[] Base =
    [
        VisualPropertyKeys.X,
        VisualPropertyKeys.Y,
        VisualPropertyKeys.Width,
        VisualPropertyKeys.Height,
        VisualPropertyKeys.ZIndex,
        VisualPropertyKeys.Rotation,
        VisualPropertyKeys.ScaleX,
        VisualPropertyKeys.ScaleY,
        VisualPropertyKeys.HorizontalFlip,
        VisualPropertyKeys.VerticalFlip,
        VisualPropertyKeys.Visible,
        VisualPropertyKeys.Opacity,
        VisualPropertyKeys.Tooltip,
        VisualPropertyKeys.Enabled,
        VisualPropertyKeys.ShadowEnabled,
        VisualPropertyKeys.ShadowColor,
        VisualPropertyKeys.ShadowOffsetX,
        VisualPropertyKeys.ShadowOffsetY,
        VisualPropertyKeys.ShadowBlur
    ];

    private static readonly string[] Fill =
    [
        VisualPropertyKeys.FillStyle,
        VisualPropertyKeys.FillColor,
        VisualPropertyKeys.FillSecondaryColor,
        VisualPropertyKeys.GradientDirection
    ];

    private static readonly string[] Stroke =
    [
        VisualPropertyKeys.StrokeColor,
        VisualPropertyKeys.StrokeWidth,
        VisualPropertyKeys.StrokeStyle
    ];

    private static readonly string[] TextProperties =
    [
        VisualPropertyKeys.Text,
        VisualPropertyKeys.TextColor,
        VisualPropertyKeys.FontFamily,
        VisualPropertyKeys.FontSize,
        VisualPropertyKeys.FontWeight,
        VisualPropertyKeys.FontStyle,
        VisualPropertyKeys.Underline,
        VisualPropertyKeys.TextWrap,
        VisualPropertyKeys.LineHeight,
        VisualPropertyKeys.TextOverflow,
        VisualPropertyKeys.HorizontalAlignment,
        VisualPropertyKeys.VerticalAlignment
    ];

    public static VisualObjectPropertySchema Group { get; } = Create(GroupType, Base);

    public static VisualObjectPropertySchema Rectangle { get; } = Create(
        RectangleType,
        Base
            .Concat(Fill)
            .Concat(Stroke)
            .Concat([VisualPropertyKeys.CornerRadius]));

    public static VisualObjectPropertySchema Ellipse { get; } = Create(
        EllipseType,
        Base
            .Concat(Fill)
            .Concat(Stroke));

    public static VisualObjectPropertySchema Line { get; } = Create(
        LineType,
        Base.Concat(Stroke));

    /// <summary>
    /// Polygon points are structural geometry owned by the core.polygon contract,
    /// not a scalar Visual Property Registry value. Only common transform/
    /// appearance properties belong to the shared property schema.
    /// </summary>
    public static VisualObjectPropertySchema Polygon { get; } = Create(
        PolygonType,
        Base
            .Concat(Fill)
            .Concat(Stroke));

    public static VisualObjectPropertySchema Text { get; } = Create(
        TextType,
        Base.Concat(TextProperties));

    public static VisualObjectPropertySchema Image { get; } = Create(
        ImageType,
        Base.Concat(
        [
            VisualPropertyKeys.AssetRef,
            VisualPropertyKeys.ImageFit,
            VisualPropertyKeys.ImagePositionX,
            VisualPropertyKeys.ImagePositionY
        ]));

    public static VisualObjectPropertySchema ValueDisplay { get; } = Create(
        ValueDisplayType,
        Base
            .Concat([VisualPropertyKeys.BackgroundColor])
            .Concat(Stroke)
            .Concat([VisualPropertyKeys.CornerRadius])
            .Concat(TextProperties));

    /// <summary>
    /// Trend pens are structural payload owned by core.trend and are deliberately
    /// excluded from the scalar Visual Property Registry. The scalar contract is
    /// kept in lockstep with the browser canonical registry.
    /// </summary>
    public static VisualObjectPropertySchema Trend { get; } = Create(
        TrendType,
        Base.Concat(
        [
            VisualPropertyKeys.BackgroundColor,
            VisualPropertyKeys.StrokeColor,
            VisualPropertyKeys.StrokeWidth,
            VisualPropertyKeys.CornerRadius,
            TrendModeProperty,
            TrendWindowSecondsProperty,
            TrendRefreshSecondsProperty,
            TrendLegendVisibleProperty,
            TrendGridVisibleProperty,
            TrendAxesVisibleProperty,
            TrendQualityVisibleProperty
        ]));

    public static VisualObjectPropertySchema Button { get; } = Create(
        ButtonType,
        Base
            .Concat([VisualPropertyKeys.BackgroundColor])
            .Concat(Stroke)
            .Concat([VisualPropertyKeys.CornerRadius])
            .Concat(TextProperties));

    public static VisualObjectPropertySchema Slider { get; } = Create(
        SliderType,
        Base.Concat(
        [
            VisualPropertyKeys.Value,
            VisualPropertyKeys.Minimum,
            VisualPropertyKeys.Maximum,
            VisualPropertyKeys.Step,
            VisualPropertyKeys.Orientation,
            VisualPropertyKeys.InteractionEnabled,
            VisualPropertyKeys.ReverseDirection,
            VisualPropertyKeys.TrackColor,
            VisualPropertyKeys.ThumbColor,
            VisualPropertyKeys.StrokeColor,
            VisualPropertyKeys.StrokeWidth,
            VisualPropertyKeys.CornerRadius
        ]));

    public static IReadOnlyCollection<VisualObjectPropertySchema> All { get; } =
    [
        Group,
        Rectangle,
        Ellipse,
        Line,
        Polygon,
        Text,
        Image,
        ValueDisplay,
        Trend,
        Button,
        Slider
    ];

    /// <summary>
    /// Public object-capability declaration for FOLLOW-B Analog Fill. Eligibility
    /// is explicit and renderer-independent; renderers must not infer it from CSS,
    /// geometry implementation or the mere presence of a color property.
    /// </summary>
    public static bool SupportsAnalogFill(string objectType) =>
        !string.IsNullOrWhiteSpace(objectType) && AnalogFillCapableTypes.Contains(objectType);

    public static VisualObjectPropertySchema GetRequired(string objectType)
    {
        var schema = All.SingleOrDefault(
            candidate => candidate.ObjectTypeKey.Equals(objectType, StringComparison.Ordinal));
        return schema ?? throw new KeyNotFoundException(
            $"Built-in visual object type '{objectType}' is not registered.");
    }

    private static VisualObjectPropertySchema Create(
        string objectType,
        IEnumerable<string> propertyKeys)
    {
        var builder = new VisualPropertySchemaBuilder(objectType);
        foreach (var propertyKey in propertyKeys)
        {
            if (!CommonByKey.TryGetValue(propertyKey, out var definition))
                throw new InvalidOperationException(
                    $"Built-in object type '{objectType}' references unknown visual property '{propertyKey}'.");
            builder.Add(definition);
        }
        return builder.Build();
    }
}
